src/app/backend/handler/terminal.go

```diff
@@ -0,0 +1,264 @@
+  if terminalSession, ok = terminalSessions[msg.SessionID]; !ok {
+    log.Printf("handleTerminalSession: can't find session '%s'", msg.SessionID)
+    return
+  }
+
+  terminalSession.sockJSSession = session
+  terminalSession.bound <- nil
+  terminalSessions[msg.SessionID] = terminalSession
+}
+
+// Called from main for /api/sockjs
+func CreateAttachHandler(path string) (http.Handler) {
+  options := sockjs.DefaultOptions
+  // options.Websocket = false
+  return sockjs.NewHandler(path, options, handleTerminalSession)
+}
+
+// Called by handleAttach
+// Executed cmd in the container specified in request and connects it up with the ptyHandler (a session)
+func execHelper(request *restful.Request, cmd []string, ptyHandler PtyHandler) error {
+	namespace := request.PathParameter("namespace")
+	podName := request.PathParameter("pod")
+	containerName := request.PathParameter("container")
+
+	req := getApiClient(request).Core().RESTClient().Post().
+		Resource("pods").
+		Name(podName).
+		Namespace(namespace).
+		SubResource("exec")
+
+	req.VersionedParams(&api.PodExecOptions{
+		Container: containerName,
+		Command:   cmd,
+		Stdin:     true,
+		Stdout:    true,
+		Stderr:    true,
+		TTY:       true,
+	}, api.ParameterCodec)
+
+	exec, err := remotecommand.NewExecutor(getApiConfig(request), "POST", req.URL())
+	if err != nil {
+		return err
+	}
+
+	err = exec.Stream(remotecommand.StreamOptions{
+		SupportedProtocols: remotecommandconsts.SupportedStreamingProtocols,
+		Stdin:              ptyHandler,
+		Stdout:             ptyHandler,
+		Stderr:             ptyHandler,
+		TerminalSizeQueue:  ptyHandler,
+		Tty:                true,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Generates a random session ID string. The format is not really interesting.
+// This ID is used to identify the session when the client opens the SockJS connection.
+// Not the same as the SockJS session id! We can't use that as that is generated
+// on the client side and we don't have it yet at this point.
+func genTerminalSessionId() (string, error) {
+	bytes := make([]byte, 16)
+	if _, err := rand.Read(bytes); err != nil {
+		return "", err
+	}
+	id := make([]byte, hex.EncodedLen(len(bytes)))
+	hex.Encode(id, bytes)
+	return string(id), nil
+}
+
+// Checks if the shell is an allowed one
+func isValidShell(validShells, shell []string) bool {
+	if len(shell) == 1 {
+		for _, validShell := range validShells {
+			if validShell == shell[0] {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// Called from apihandler.handleAttach as a goroutine
+// Waits for the SockJS connection to be opened by the client the session to be bound in handleTerminalSession
+func TerminalWaiter(request *restful.Request, sessionId string) {
+	shell := []string{request.QueryParameter("shell")}
+
+	select {
+	case <-terminalSessions[sessionId].bound:
+		close(terminalSessions[sessionId].bound)
+
+		var err error
+		validShells := []string{"bash", "sh"}
+
+		if shell[0] != "" {
```

Comment:
Instead of doing this check check if `isValidShell(validShells, shell)`, else `try some shells until one succeeds or all fail`. Then, we don't need to throw `err = fmt.Errorf("Invalid shell")`.